geode-core/src/test/java/org/apache/geode/cache/query/internal/MethodDispatchTest.java

```diff
@@ -62,20 +70,180 @@ public void invokeShouldThrowExceptionWhenMethodIsNotAuthorized() throws NameRes
+      assertThatThrownBy(
+          () -> anotherPublicMethodDispatch.invoke(testBean, emptyList, queryExecutionContext))
+              .isInstanceOf(NotAuthorizedException.class)
+              .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "anotherPublicMethod");
+      assertThatThrownBy(() -> extendedPublicMethodDispatch
+          .invoke(new TestBeanExtension(), emptyList, queryExecutionContext))
+              .isInstanceOf(NotAuthorizedException.class)
+              .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "publicMethod");
+    });
+    verify(methodInvocationAuthorizer, times(3)).authorize(any(), any());
+  }
+
+  @Test
+  public void invokeShouldNotUseAlreadyAuthorizedCachedResultWhenMethodIsNotAuthorizedAndContextIsNotTheSame()
+      throws NameResolutionException {
+    doReturn(false).when(methodInvocationAuthorizer).authorize(any(), any());
+    MethodDispatch publicMethodDispatch =
+        new MethodDispatch(methodInvocationAuthorizer, TestBean.class, "publicMethod",
+            Collections.emptyList());
+    MethodDispatch anotherPublicMethodDispatch =
+        new MethodDispatch(methodInvocationAuthorizer, TestBean.class, "anotherPublicMethod",
+            Collections.emptyList());
+    MethodDispatch extendedPublicMethodDispatch =
+        new MethodDispatch(methodInvocationAuthorizer, TestBeanExtension.class, "publicMethod",
+            Collections.emptyList());
+
+    // New QueryExecutionContext every time -> Do not use cache.
+    IntStream.range(0, 10).forEach(element -> {
+      QueryExecutionContext mockContext = mock(QueryExecutionContext.class);
+      assertThatThrownBy(() -> publicMethodDispatch.invoke(testBean, emptyList, mockContext))
+          .isInstanceOf(NotAuthorizedException.class)
+          .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "publicMethod");
+      assertThatThrownBy(() -> anotherPublicMethodDispatch.invoke(testBean, emptyList, mockContext))
+          .isInstanceOf(NotAuthorizedException.class)
+          .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "anotherPublicMethod");
+      assertThatThrownBy(() -> extendedPublicMethodDispatch
+          .invoke(new TestBeanExtension(), emptyList, mockContext))
+              .isInstanceOf(NotAuthorizedException.class)
+              .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "publicMethod");
+    });
+    verify(methodInvocationAuthorizer, times(30)).authorize(any(), any());
+  }
+
   @SuppressWarnings("unused")
   private static class TestBean {
-
     public String publicMethod() {
       return "publicMethod";
     }
 
-    private String privateMethod() {
-      return "privateMethod";
+    public String anotherPublicMethod() {
+      return "anotherPublicMethod";
+    }
+  }
+
+  private static class TestBeanExtension extends TestBean {
+    @Override
+    public String publicMethod() {
+      return "extendedPublicMethod";
```

Comment:
I think these strings should be extracted to constants as they're referred to repeatedly in tests. Also, maybe have constants to refer to the method names to avoid confusing between the string referring to the method name and the string that is returned by the method.

Reply:
👍